### PR TITLE
Modernize crate

### PR DESCRIPTION
**Pull request by Emilgardis**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -209,7 +209,7 @@ linker = "rust-lld"
 
 ## Minimum Supported Rust Version (MSRV)
 
-This crate is guaranteed to compile on stable Rust 1.60.0 and up. It *might*
+This crate is guaranteed to compile on stable Rust 1.64.0 and up. It *might*
 compile with older versions but that may change in any new patch release.
```
Comment (Emilgardis) [inline README.md:213]: this `*might*` is now irrelevant (with `rust-version`), we could consider removing this sentence

File: src/lib.rs [hunk 2]
```diff
@@ -1,5 +1,3 @@
-#![deny(warnings)]
-
```
Comment (Emilgardis) [inline src/lib.rs:2]: deny in ci instead

File: src/lib.rs [hunk 3]
```diff
@@ -159,119 +157,111 @@ To see all the flags the proxied tool accepts run `cargo-{} -- --help`.{}",
...
                 .help("Build artifacts with the specified profile"),
-            Arg::with_name("features")
+            Arg::new("features")
                 .long("features")
-                .multiple(true)
-                .number_of_values(1)
-                .takes_value(true)
+                .action(ArgAction::Set)
                 .value_name("FEATURES")
                 .help("Space-separated list of features to activate"),
```
Comment (Emilgardis) [inline src/lib.rs:234]: im not super confident about this, why did we specify multiple here?

anyway, this is wrong, should be `ArgAction::Append`

File: src/lib.rs [hunk 4]
```diff
@@ -159,119 +157,111 @@ To see all the flags the proxied tool accepts run `cargo-{} -- --help`.{}",
...
                 .long("offline")
                 .help("Run without accessing the network"),
-            Arg::with_name("unstable-features")
-                .short("Z")
-                .multiple(true)
-                .number_of_values(1)
-                .takes_value(true)
+            Arg::new("unstable-features")
+                .short('Z')
+                .action(ArgAction::Set)
```
Comment (Emilgardis) [inline src/lib.rs:264]: this is wrong, should be `ArgAction::Append`

File: src/lib.rs [hunk 5]
```diff
@@ -159,119 +157,106 @@ To see all the flags the proxied tool accepts run `cargo-{} -- --help`.{}",
...
                 .long("profile")
                 .value_name("PROFILE-NAME")
                 .help("Build artifacts with the specified profile"),
-            Arg::with_name("features")
+            Arg::new("features")
                 .long("features")
-                .multiple(true)
-                .number_of_values(1)
-                .takes_value(true)
+                .short('F')
```
Comment (Emilgardis) [inline src/lib.rs:228]: a little bonus, `-F` was added in 1.62.0

**Review by burrbull — APPROVED** [review]

**Comment by Emilgardis** [comment]
Can we get some more eyes on this, don't want to accidentally break anything

**Comment by Emilgardis** [comment]
ping @rust-embedded/tools

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -24,51 +24,38 @@ jobs:
             experimental: true
 
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions-rs/toolchain@v1
+      - uses: actions/checkout@v3
+      - uses: dtolnay/rust-toolchain@9cd00a88a73addc8617065438eff914dd08d0955 # 2023-02-02
```
Comment (thejpster) [inline .github/workflows/ci.yml:28]: Why use this action? And why is it at a git hash not a version?

Comment (Emilgardis) [reply .github/workflows/ci.yml:28]: The author has expressed they dont want to provide tags except for when its used to specify rust versions. Afaik, `uses` can not be evaluated like `uses: ${{ matrix.var }}`

Comment (burrbull) [reply .github/workflows/ci.yml:28]: checkout@v4

**Comment by Emilgardis** [comment]
I'm confused, is actions/checkout@v4 now merging `pull_request` head with `base`? https://github.com/rust-embedded/cargo-binutils/actions/runs/6265523508/job/17014546817?pr=132#step:2:77

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -19,20 +19,12 @@ jobs:
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v3
-      - uses: actions-rs/toolchain@v1
+      - uses: dtolnay/rust-toolchain@9cd00a88a73addc8617065438eff914dd08d0955 # 2023-02-02
```
Comment (thejpster) [inline .github/workflows/release.yml:22]: Same comment as CI.yml...

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,8 +9,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Changed
 
-- MSRV Changed to 1.60
-- Bump `rust-cfg` to 0.5 and `cargo_metadata` to 0.15
+- MSRV Changed to 1.64
```
Comment (thejpster) [inline CHANGELOG.md:12]: The build matrix was set to 1.72.1? What is the MSRV supposed to be?

Comment (Emilgardis) [reply CHANGELOG.md:12]: Missed this one

**Comment by jonathanpallant** [comment]
What's the reason for moving the MSRV to 1.72.1? Debian [is shipping 1.70](https://packages.debian.org/sid/rustc).

**Comment by Emilgardis** [comment]
The reason is because we can, and we dont have a policy except that we may bump the msrv whenever according to https://github.com/rust-embedded/wg/blob/master/rfcs/0523-msrv-2020.md

**Comment by thejpster** [comment]
What is the lowest Rust version that actually work for this crate? i.e. what actually is our *minimum* supported Rust Version at this time?

**Comment by Emilgardis** [comment]
clap is the limiter pretty sure, so 1.70

**Comment by therealprof** [comment]
I think we shouldn't be too agressive with moving the MSRV. It's certainly worthwile to be a bit conservative here. Also if we already know that the bump of `clap` requires 1.70 we shouldn't be mentioning that it might build with 1.64.

Unfortunately I've made the experience that between clap 2 and clap 3 there had been a ton of oh so subtle changes which broke some uses of my tools and took quite some time to discover and fix since the authors decided that it was undocumented behaviour and thus fair to silently change -- I'm a bit worried that we might run into similar issues. At the very least we would have a big fat warning in a few places to alert people to the possibility of unintended CLI breakage due to the `clap` bump.

**Comment by Emilgardis** [comment]
we could also publish a lockfile, and then recommend installation with `--locked`, also #138 would mitigate this a bit since we prepare the binary

**Review by burrbull — APPROVED** [review]

**Comment by burrbull** [comment]
need to migrate on `merge queue`

**Review by burrbull — APPROVED** [review]

**Pull request merged**